### PR TITLE
Chore: Resolve Remaining Rubocop RSpec Todo's

### PR DESCRIPTION
**Pull request by KevinMulhern**
Because:
* We want all rubocop rspec violations to be fixed.

This commit
* Refactor lesson completions api controller spec to a request spec
* Fix RSpec/MultipleMemoizedHelpers violations

closes https://github.com/TheOdinProject/theodinproject/issues/2981

Complete the following REQUIRED checkboxes:

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `keyword: brief description of change` format, using one of the following keywords:
  - `Feature` - adds new or amends existing user-facing behaviour
  - `Chore` - changes that have no user-facing value, refactors, dependency bumps, etc
  - `Fix` - bug fixes

Complete the following checkbox ONLY IF it is applicable to your PR. You can complete it later if it is not currently applicable:
-   [x] I have verified all tests and linters pass against my changes, and/or I have included automated tests where applicable

### CODE REVIEW

File: spec/requests/api/lesson_completions_spec.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                  'title' => lesson.title,
+                  'completions' => 2
+                ]
+              }
+            ]
+          )
+        )
+      end
+    end
+
```
Comment (dm-murphy) [inline spec/requests/api/lesson_completions_spec.rb:98]: **question (non-blocking):**  Just checking, but do we no longer need the test for missing start and end dates from the old spec?

Original controller spec:
`context 'when start and end dates are not present' do`
`it 'uses the default dates' do`

File: spec/services/messages/dead_link_spec.rb [hunk 2]
```diff
@@ -7,15 +7,11 @@
...
       created_at: Time.zone.local(2021, 8, 1),
-      id: 120,
-      flagger:,
-      project_submission: flagged_submission,
+      project_submission: create(:project_submission, lesson:, user:),
       reason: 'I find it offensive'
     )
   end
 
-  let(:flagger) { create(:user, username: 'OdinUser') }
```
Comment (dm-murphy) [inline spec/services/messages/dead_link_spec.rb:17]: **question (non-blocking):** Is `flagger` no longer needed at all? I just didn't understand this one fully 😄

**Review by dm-murphy — APPROVED** [review]
Awesome work @KevinMulhern 💪 

Thanks for changing the controller spec into a request and for fixing those remaining rubocop issues :-)

Approved with a couple comments left to help me try and learn 😆

File: spec/requests/api/lesson_completions_spec.rb (hunk 1 above)
Comment (KevinMulhern) [reply spec/requests/api/lesson_completions_spec.rb:121]: Thanks @dm-murphy, Great spot 🦅 👁️

File: spec/services/messages/dead_link_spec.rb (hunk 2 above)
Comment (KevinMulhern) [reply spec/services/messages/dead_link_spec.rb:17]: Yeah flagger was only used for setting up the flag factory and nowhere else in the test. We can get rid of it with no downsides by relying on the flagger association set up by default when a flagger isn't provided.

Comment (dm-murphy) [reply spec/services/messages/dead_link_spec.rb:17]: Gotcha, thanks for clarifying :-)

File: spec/requests/api/lesson_completions_spec.rb (hunk 1 above)
Comment (dm-murphy) [reply spec/requests/api/lesson_completions_spec.rb:121]: Thanks! Happy to help!

**Review by dm-murphy — APPROVED** [review]

**Pull request merged**